test/integration/mhc_test.go

```diff
@@ -0,0 +1,253 @@
+// +build integration
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"flag"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	"k8s.io/klog/klogr"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers"
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	"sigs.k8s.io/cluster-api/test/helpers"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+func TestMachineSetsWithMachineHealthCheck(t *testing.T) {
+	g := NewWithT(t)
+	testEnv, err := helpers.NewTestEnvironment()
+	g.Expect(err).NotTo(HaveOccurred())
+	defer func() {
+		g.Expect(testEnv.Stop()).To(Succeed())
+	}()
+
+	klog.InitFlags(nil)
+	g.Expect(flag.Set("v", "4")).To(Succeed())
+	g.Expect(flag.Set("logtostderr", "false")).To(Succeed())
+
+	logger := klogr.New()
+
+	g.Expect((&controllers.MachineSetReconciler{
+		Client: testEnv,
+		Log:    logger,
+	}).SetupWithManager(testEnv.Manager, controller.Options{MaxConcurrentReconciles: 1})).To(Succeed())
+	g.Expect((&controllers.MachineHealthCheckReconciler{
+		Client: testEnv,
+		Log:    logger,
+	}).SetupWithManager(testEnv.Manager, controller.Options{MaxConcurrentReconciles: 1})).To(Succeed())
+
+	go func() {
+		g.Expect(testEnv.StartManager()).To(Succeed())
+	}()
```

Comment:
Can't we reuse the same suite / envtest that we have under `controllers/`?

Reply:
I think we should have these kinds of integration tests that can span across multiple controllers, where those ones are just tests of the individual controllers. This one exercises both the MachineHealthCheck reconciler as well as the MachineSet reconciler to show that they are wired up together correctly. I also think the ones that we have in `controllers` should be unit tests and shouldn't be using envtest.

Reply:
The envtest under controller runs all controllers at the same time, we should be able to achieve the same we're doing here without changes.